File: testing/adios2/bindings/fortran/TestBPWriteTypesHighLevelAPI.f90
 program TestBPWriteTypesHighLevelAPI
   use small_test_data
   use mpi
   use adios2
   implicit none

   ! high-level writer
   type(adios2_file) :: fhw, fhr
   ! low-level reader
   type(adios2_adios) :: adios
   type(adios2_io) :: ioRead
   type(adios2_variable), dimension(12) :: variables
   type(adios2_engine) :: bpReader

   integer(kind=8), dimension(1) :: shape_dims, start_dims, count_dims
   integer :: inx, irank, isize, ierr, i, n

   ! read handlers
   character(len=:), allocatable :: variable_name
   integer :: variable_type, ndims
   integer(kind=8), dimension(:), allocatable :: shape_in

   ! read arrays
   integer(kind=1) :: in_gi8
   integer(kind=2) :: in_gi16
   integer(kind=4) :: in_gi32
   integer(kind=8) :: in_gi64
   real(kind=4)    :: in_gr32
   real(kind=8)    :: in_gr64

   integer(kind=1), dimension(:), allocatable :: in_I8
   integer(kind=2), dimension(:), allocatable :: in_I16
   integer(kind=4), dimension(:), allocatable :: in_I32
   integer(kind=8), dimension(:), allocatable :: in_I64
   real, dimension(:), allocatable :: in_R32
   real(kind=8), dimension(:), allocatable :: in_R64

   ! Launch MPI
   call MPI_Init(ierr)
   call MPI_Comm_rank(MPI_COMM_WORLD, irank, ierr)
   call MPI_Comm_size(MPI_COMM_WORLD, isize, ierr)

   ! Application variables
   inx = 10

   ! Variable dimensions
   shape_dims(1) = isize*inx
   start_dims(1) = irank*inx
   count_dims(1) = inx

   !!!!!!!!!!!!!!!!!!!!!!!!! WRITER !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
   call adios2_fopen(fhw, 'ftypes_hl.bp', adios2_mode_write, &
                     MPI_COMM_WORLD, ierr)

   ! Global single value variables
   if (irank == 0) then
     call adios2_fwrite(fhw, "gvar_I8", data_I8(1), ierr)
     call adios2_fwrite(fhw, "gvar_I16", data_I16(1), ierr)
     call adios2_fwrite(fhw, "gvar_I32", data_I32(1), ierr)
     call adios2_fwrite(fhw, "gvar_I64", data_I64(1), ierr)
     call adios2_fwrite(fhw, "gvar_R32", data_R32(1), ierr)
     call adios2_fwrite(fhw, "gvar_R64", data_R64(1), ierr)
   end if

   ! 1D array variables
   do i = 1, 3
     call adios2_fwrite(fhw, "var_I8", data_I8, 1, shape_dims, &
                        start_dims, count_dims, ierr)

     call adios2_fwrite(fhw, "var_I16", data_I16, 1, shape_dims, &
                        start_dims, count_dims, ierr)

     call adios2_fwrite(fhw, "var_I32", data_I32, 1, shape_dims, &
                        start_dims, count_dims, ierr)

     call adios2_fwrite(fhw, "var_I64", data_I64, 1, shape_dims, &
                        start_dims, count_dims, ierr)

     call adios2_fwrite(fhw, "var_R32", data_R32, 1, shape_dims, &
                        start_dims, count_dims, ierr)

     call adios2_fwrite(fhw, "var_R64", data_R64, 1, shape_dims, &
                        start_dims, count_dims, ADIOS2_ADVANCE_YES, ierr)
   end do

   call adios2_fclose(fhw, ierr)

   !!!!!!!!!!!!!!!!!!!!!!!!! High-level READER !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
   call adios2_fopen(fhr, 'ftypes_hl.bp', adios2_mode_read, &
                     MPI_COMM_WORLD, ierr)

   i = 0
   do

     if (i == 0) then
       call adios2_fread(fhr, "gvar_I8", in_gI8, ierr)
       write (*, *) 'in_gI8 ', in_gI8
       if (in_gI8 /= data_I8(1)) stop "in_I8 not equal data_I8(1)"

       call adios2_fread(fhr, "gvar_I16", in_gI16, ierr)
       write (*, *) 'in_gI16 ', in_gI16
       if (in_gI16 /= data_I16(1)) stop "in_gI16 not equal data_I16(1)"

       call adios2_fread(fhr, "gvar_I32", in_gI32, ierr)
       if (in_gI32 /= data_I32(1)) stop "in_gI32 not equal data_I32(1)"

       call adios2_fread(fhr, "gvar_I64", in_gI64, ierr)
       if (in_gI64 /= data_I64(1)) stop "in_gI64 not equal data_I64(1)"

       call adios2_fread(fhr, "gvar_R32", in_gR32, ierr)
       if (in_gR32 /= data_R32(1)) stop "in_gR32 not equal data_R32(1)"

       call adios2_fread(fhr, "gvar_R64", in_gR64, ierr)
       if (in_gR64 /= data_R64(1)) stop "in_gR64 not equal data_R64(1)"
     end if

     call adios2_fread(fhr, "var_I8", in_I8, start_dims, count_dims, ierr)
     call adios2_fread(fhr, "var_I16", in_I16, start_dims, count_dims, ierr)
     call adios2_fread(fhr, "var_I32", in_I32, start_dims, count_dims, ierr)
     call adios2_fread(fhr, "var_I64", in_I64, start_dims, count_dims, ierr)
     call adios2_fread(fhr, "var_R32", in_R32, start_dims, count_dims, ierr)
     call adios2_fread(fhr, "var_R64", in_R64, start_dims, count_dims, &
                       ADIOS2_ADVANCE_YES, ierr)

     do n = 1, count_dims(1)
       if (in_I8(n) /= data_I8(n)) stop 'I8 read failure'
       if (in_I16(n) /= data_I16(n)) stop 'I16 read failure'
       if (in_I32(n) /= data_I32(n)) stop 'I32 read failure'
       if (in_I64(n) /= data_I64(n)) stop 'I64 read failure'
       if (in_R32(n) /= data_R32(n)) stop 'R32 read failure'
       if (in_R64(n) /= data_R64(n)) stop 'R64 read failure'
     end do

     i = i + 1
     if (ierr < 0) exit

   end do

   call adios2_fclose(fhr, ierr)

   if (allocated(in_I8)) deallocate (in_I8)
   if (allocated(in_I16)) deallocate (in_I16)
   if (allocated(in_I32)) deallocate (in_I32)
   if (allocated(in_I64)) deallocate (in_I64)
   if (allocated(in_R32)) deallocate (in_R32)
   if (allocated(in_R64)) deallocate (in_R64)

   call MPI_Finalize(ierr)

 end program TestBPWriteTypesHighLevelAPI

File: testing/adios2/bindings/python/TestBPWriteTypesHighLevelAPI.py
#!/usr/bin/env python

#
# Distributed under the OSI-approved Apache License, Version 2.0.  See
# accompanying file Copyright.txt for details.
#
# TestBPWriteTypes.py: test Python numpy types in ADIOS2 File
#                      Write/Read High-Level API
#  Created on: March 12, 2018
#      Author: William F Godoy godoywf@ornl.gov


from adios2NPTypes import SmallTestData
from mpi4py import MPI
import numpy as np
import adios2

comm = MPI.COMM_WORLD
rank = comm.Get_rank()
size = comm.Get_size()

# Test data
data = SmallTestData()
nx = data.Nx

shape = [size * nx]
start = [rank * nx]
count = [nx]

# Writer
fw = adios2.open("types_np.bp", "w", comm)


if(rank == 0):
    fw.write("tag", "Testing ADIOS2 high-level API")
    fw.write("gvarI8", np.array(data.I8[0]))
    fw.write("gvarI16", np.array(data.I16[0]))
    fw.write("gvarI32", np.array(data.I32[0]))
    fw.write("gvarI64", np.array(data.I64[0]))
    fw.write("gvarU8", np.array(data.U8[0]))
    fw.write("gvarU16", np.array(data.U16[0]))
    fw.write("gvarU32", np.array(data.U32[0]))
    fw.write("gvarU64", np.array(data.U64[0]))
    fw.write("gvarR32", np.array(data.R32[0]))
    fw.write("gvarR64", np.array(data.R64[0]))

for i in range(0, 3):
    fw.write("steps", "Step:" + str(i))

    fw.write("varI8", data.I8, shape, start, count)
    fw.write("varI16", data.I16, shape, start, count)
    fw.write("varI32", data.I32, shape, start, count)
    fw.write("varI64", data.I64, shape, start, count)
    fw.write("varU8", data.U8, shape, start, count)
    fw.write("varU16", data.U16, shape, start, count)
    fw.write("varU32", data.U32, shape, start, count)
    fw.write("varU64", data.U64, shape, start, count)
    fw.write("varR32", data.R32, shape, start, count)
    fw.write("varR64", data.R64, shape, start, count, True)

fw.close()

# Reader
with adios2.open("types_np.bp", "r", comm) as fr:
 
    step_vars = fr.availablevariables()
         
    for name, info in step_vars.items():
        print("variable_name: " + name)
        for key, value in info.items():
            print("\t" + key + ": " + value)
        print("\n")
    
    
    for step in fr:
    
        print("Step" + str(step))
        step_vars = fr.availablevariables()
         
        for name, info in step_vars.items():
            print("variable_name: " + name)
            for key, value in info.items():
                print("\t" + key + ": " + value)
            print("\n")
        
        if(fr.currenstep() == 1):
            inTag = fr.readstring("tag")
            inI8 = fr.read("gvarI8")
            inI16 = fr.read("gvarI16")
            inI32 = fr.read("gvarI32")
            inI64 = fr.read("gvarI64")
            inU8 = fr.read("gvarU8")
            inU16 = fr.read("gvarU16")
            inU32 = fr.read("gvarU32")
            inU64 = fr.read("gvarU64")
            inR32 = fr.read("gvarR32")
            inR64 = fr.read("gvarR64")
    
            if(inTag != "Testing ADIOS2 high-level API"):
                raise ValueError('tag read failed')
    
            if(inI8[0] != data.I8[0]):
                raise ValueError('gvarI8 read failed')
    
            if(inI16[0] != data.I16[0]):
                raise ValueError('gvarI16 read failed')
    
            if(inI32[0] != data.I32[0]):
                raise ValueError('gvarI32 read failed')
    
            if(inU16[0] != data.U16[0]):
                raise ValueError('gvarU16 read failed')
    
            if(inI64[0] != data.I64[0]):
                raise ValueError('gvarI64 read failed')
    
            if(inU64[0] != data.U64[0]):
                raise ValueError('gvarU64 read failed')
    
            if(inI8[0] != data.I8[0]):
                raise ValueError('gvarI8 read failed')
            
            if(inU8[0] != data.U8[0]):
                raise ValueError('gvarU8 read failed')
            
            if(inU32[0] != data.U32[0]):
                raise ValueError('gvarU32 read failed')
            
            if(inR32[0] != data.R32[0]):
                raise ValueError('gvarR32 read failed')
            
            if(inR64[0] != data.R64[0]):
                raise ValueError('gvarR64 read failed')
    
        instepStr = fr.readstring("steps")
        indataI8 = fr.read("varI8", start, count)
        indataI16 = fr.read("varI16", start, count)
        indataI32 = fr.read("varI32", start, count)
        indataI64 = fr.read("varI64", start, count)
        indataU8 = fr.read("varU8", start, count)
        indataU16 = fr.read("varU16", start, count)
        indataU32 = fr.read("varU32", start, count)
        indataU64 = fr.read("varU64", start, count)
        indataR32 = fr.read("varR32", start, count)
        indataR64 = fr.read("varR64", start, count)
    
        stepStr = "Step:" + str(i).strip()
        i = i + 1
    
        if(instepStr != stepStr):
            raise ValueError(
                'steps variable read failed: ' + instepStr + " " + stepStr)
    
        if((indataI8 == data.I8).all() is False):
            print("InData: " + str(indataI8))
            print("Data: " + str(data.I8))
            raise ValueError('I8 array read failed')
    
        if((indataI16 == data.I16).all() is False):
            print("InData: " + str(indataI16))
            print("Data: " + str(data.I16))
            raise ValueError('I16 array read failed')
    
        if((indataI32 == data.I32).all() is False):
            print("InData: " + str(indataI32))
            print("Data: " + str(data.I32))
            raise ValueError('I32 array read failed')
    
        if((indataI64 == data.I64).all() is False):
            print("InData: " + str(indataI64))
            print("Data: " + str(data.I64))
            raise ValueError('I64 array read failed')
    
        if((indataU8 == data.U8).all() is False):
            print("InData: " + str(indataU8))
            print("Data: " + str(data.U8))
            raise ValueError('U8 array read failed')
    
        if((indataU16 == data.U16).all() is False):
            print("InData: " + str(indataU16))
            print("Data: " + str(data.U16))
            raise ValueError('U16 array read failed')
    
        if((indataU32 == data.U32).all() is False):
            print("InData: " + str(indataU32))
            print("Data: " + str(data.U32))
            raise ValueError('U32 array read failed')
    
        if((indataU64 == data.U64).all() is False):
            print("InData: " + str(indataU64))
            print("Data: " + str(data.U64))
            raise ValueError('U64 array read failed')
    
        if((indataR32 == data.R32).all() is False):
            print("InData: " + str(indataR32))
            print("Data: " + str(data.R32))
            raise ValueError('R32 array read failed')
    
        if((indataR64 == data.R64).all() is False):
            print("InData: " + str(indataR64))
            print("Data: " + str(data.R64))
            raise ValueError('R64 array read failed')
